Create migration for removing orphaned user data

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-16810

This can't actually be merged until the underlying bug is fixed (https://github.com/dimagi/commcare-hq/pull/36000) which is to delete user data when removing a user from a domain. However I wanted to get this PR up to run this for a specific domain to unblock a test I'm conducting.

Now that https://github.com/dimagi/commcare-hq/pull/36000 is merged, this is ready for review. The gist of this is a management command that grabs all user data for a domain, and checks if the `user_id` for that user data either no longer exists, or does not have a domain membership. If either of those is true, the user data is considered orphaned and therefore deleted.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
In addition to the automated test coverage, I've run this on a specific domain once it was approved initially and it behaved as expected. If anything, this code is structured in such a way where it would not delete something it should have rather than deleting something is shouldn't, which seems good. I say this because the checks are pretty simple (if the CouchUser or DomainMembership evaluates to False, we delete user data).
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Added tests for the management command to ensure this works as expected.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Create migration for removing orphaned user data

orangejenny commented on corehq/apps/users/management/commands/rm_orphaned_user_data.py:
```diff
@@ -0,0 +1,40 @@
+            continue
+
+        if not dm:
+            if not dry_run:
```
Optional paranoia: The way you have this set up, it'd be pretty easy to reorder so that you just gather the user ids, then display them and ask the user for confirmation, hen bulk delete.

gherceg replied:
Yeah I suppose the main reason I didn't do something like this is that I envision using this command once for the domain that is blocking me, and then calling this from the migration (where there will obviously be no confirmation), and then never using it again. So leaning towards passing on this but if there is a reason I should include it, happy to.

orangejenny replied:
nah, that sounds fine

gherceg committed:
Run ./manage.py makemigrations --lock-update

gherceg committed:
Merge branch 'master' into gh/orphaned-user-data

gherceg committed:
Add proper params for migration function

gherceg committed:
Fix dry run test

Once https://github.com/dimagi/commcare-hq/pull/36000 was merged, it
made this test fail because deleting a domain membership would cleanup
user data. So I restructured the test to ensure there was orphaned user
data.

millerdev reviewed (approved):
I imagine the migration could take a while to run. Do you plan on running it in a private release before it is deployed?

How many Django users do we have (in SQL) without a corresponding Couch user? If there are any, do we know what causes the system to get into that state?

Edit: that second question was poorly phrased, and was not getting at the issue that this PR is fixing. Offlined with Graham to resolve that misunderstanding.

gherceg committed:
Merge branch 'master' into gh/orphaned-user-data

gherceg committed:
Domain argument should be optional
